File: app/src/main/java/djordjeh/architecture/mvp/data/source/local/TaskLocalDataSource.java
package djordjeh.architecture.mvp.data.source.local;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import androidx.annotation.NonNull;

import com.squareup.sqlbrite3.BriteDatabase;

import java.util.List;

import djordjeh.architecture.mvp.data.source.TaskDataSource;
import djordjeh.architecture.mvp.data.model.Task;
import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;

public class TaskLocalDataSource implements TaskDataSource {

    private final BriteDatabase database;

    public TaskLocalDataSource(BriteDatabase database) {
        this.database = database;
    }

    @Override
    public Observable<List<Task>> tasks(boolean forceUpdate) {
        final String query = String.format("SELECT * FROM %s ORDER BY %s DESC", AppSQLiteHelper.TaskEntity.TABLE_NAME, AppSQLiteHelper.TaskEntity._ID);
        return database.createQuery(AppSQLiteHelper.TaskEntity.TABLE_NAME, query).mapToList(this::getTask);
    }

    @Override
    public Maybe<Task> task(long taskId) {
        final String query = String.format("SELECT * FROM %s WHERE %s = ?", AppSQLiteHelper.TaskEntity.TABLE_NAME, AppSQLiteHelper.TaskEntity._ID);
        return database.createQuery(AppSQLiteHelper.TaskEntity.TABLE_NAME, query, String.valueOf(taskId)).mapToOne(this::getTask).firstElement();
    }

    @Override
    public Single<Task> save(@NonNull Task task) {
        final long rowId = database.insert(AppSQLiteHelper.TaskEntity.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, taskContent(task));
        if (rowId != -1) return Single.just(task);
        else return Single.error(new IllegalArgumentException("Task can't be saved"));
    }

    @Override
    public Single<Task> delete(@NonNull Task task) {
        final String whereClause = AppSQLiteHelper.TaskEntity._ID + " = ?";
        final long rowId = database.delete(AppSQLiteHelper.TaskEntity.TABLE_NAME, whereClause, String.valueOf(task.getId()));
        return rowId > 0 ? Single.just(task) : Single.error(new IllegalArgumentException("Task not find"));
    }

    private ContentValues taskContent(@NonNull Task task) {
        final ContentValues values = new ContentValues();
        values.put(AppSQLiteHelper.TaskEntity._ID, task.getId() == 0 ? System.currentTimeMillis() : task.getId());
        values.put(AppSQLiteHelper.TaskEntity.TITLE, task.getTitle());
        values.put(AppSQLiteHelper.TaskEntity.DESCRIPTION, task.getDescription());
        values.put(AppSQLiteHelper.TaskEntity.COMPLETED, task.isCompleted() ? 1 : 0);
        return values;
    }

    private Task getTask(@NonNull Cursor cursor) {
        return new Task(
                cursor.getLong(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity._ID)),
                cursor.getString(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.TITLE)),
                cursor.getString(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.DESCRIPTION)),
                cursor.getInt(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.COMPLETED)) == 1
        );
    }
}

File: app/src/main/java/djordjeh/architecture/mvp/data/source/local/TaskLocalDataSource.kt
package djordjeh.architecture.mvp.data.source.local

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase

import com.squareup.sqlbrite3.BriteDatabase

import djordjeh.architecture.mvp.data.source.TaskDataSource
import djordjeh.architecture.mvp.data.model.Task
import io.reactivex.Maybe
import io.reactivex.Observable
import io.reactivex.Single

class TaskLocalDataSource(private val database: BriteDatabase) : TaskDataSource {

    override fun tasks(forceUpdate: Boolean): Observable<List<Task>> {
        val query = String.format("SELECT * FROM %s ORDER BY %s DESC", AppSQLiteHelper.TaskEntity.TABLE_NAME, AppSQLiteHelper.TaskEntity.ID)
        return database.createQuery(AppSQLiteHelper.TaskEntity.TABLE_NAME, query).mapToList { this.getTask(it) }
    }

    override fun task(taskId: Long): Maybe<Task> {
        val query = String.format("SELECT * FROM %s WHERE %s = ?", AppSQLiteHelper.TaskEntity.TABLE_NAME, AppSQLiteHelper.TaskEntity.ID)
        return database.createQuery(AppSQLiteHelper.TaskEntity.TABLE_NAME, query, taskId.toString()).mapToOne<Task> { this.getTask(it) }.firstElement()
    }

    override fun save(task: Task): Single<Task> {
        val rowId = database.insert(AppSQLiteHelper.TaskEntity.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, taskContent(task))
        return if (rowId > 0) Single.just(task) else Single.error(IllegalArgumentException("Task can't be saved"))
    }

    override fun delete(task: Task): Single<Task> {
        val whereClause = AppSQLiteHelper.TaskEntity.ID + " = ?"
        val rowId = database.delete(AppSQLiteHelper.TaskEntity.TABLE_NAME, whereClause, task.id.toString()).toLong()
        return if (rowId > 0) Single.just(task) else Single.error(IllegalArgumentException("Task not find"))
    }

    private fun taskContent(task: Task): ContentValues {
        return ContentValues().apply {
            put(AppSQLiteHelper.TaskEntity.ID, if (task.id == 0L) System.currentTimeMillis() else task.id)
            put(AppSQLiteHelper.TaskEntity.TITLE, task.title)
            put(AppSQLiteHelper.TaskEntity.DESCRIPTION, task.description)
            put(AppSQLiteHelper.TaskEntity.COMPLETED, if (task.completed) 1 else 0)
        }
    }

    private fun getTask(cursor: Cursor): Task {
        return Task(
                cursor.getLong(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.ID)),
                cursor.getString(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.TITLE)),
                cursor.getString(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.DESCRIPTION)),
                cursor.getInt(cursor.getColumnIndex(AppSQLiteHelper.TaskEntity.COMPLETED)) == 1
        )
    }
}
